Add option to connect last changed model to AutomationClip

regulus79 opened the pull request:
This PR adds a new context menu action to `AutomationClipView`s which allows users to connect the last changed model.

A pointer to the last changed model is kept as a static variable `AutomatableModel::s_lastChangedModel`, which is updated in `AutomatableModel::setValue()`. When the `AutomationClipView` context menu is constructed, if `AutomatableModel::s_lastChangedModel` is not a nullptr and if the model's display name is not empty (Many hidden models have empty names, and it would be confusing to show them to users), then an option to connect it is given.

Demo:

https://github.com/user-attachments/assets/afcb769d-21f8-409a-9e69-60602b8f0457

sakertooth commented on src/gui/clips/AutomationClipView.cpp:
```diff
@@ -100,6 +100,23 @@ void AutomationClipView::changeName()
+							embed::getIconPixmap("automation"),
+							2000);
+		}
+		update();
```
We call `update` even in the case where the `AutomatableModel` was not added to the `AutomationClip`. Is this expected?

regulus79 replied:
I suppose it does not have to be updated if the add was not successful. I only put it there because `AutomationEditor.cpp` had a similar setup, where `update()` is called in `AutomationEditor::dropEvent` regardless of whether the model was added or not. Nevertheless, I have now reworked the if statement to only call `update()` if the add was successful.

sakertooth commented on src/gui/clips/AutomationClipView.cpp:
```diff
@@ -185,6 +202,11 @@ void AutomationClipView::constructContextMenu( QMenu * _cm )
 	_cm->addAction( embed::getIconPixmap( "flip_x" ),
 						tr( "Flip Horizontally (Visible)" ),
 						this, SLOT(flipX()));
+	if (AutomatableModel::s_lastChangedModel != nullptr && !AutomatableModel::s_lastChangedModel->displayName().isEmpty())
```
Eh, not a big fan of checking if the model name is empty and never adding the option in such a case to be honest. If it somehow turns out to be empty, that probably would mean that there is a bug elsewhere in the code that we should fix.

regulus79 replied:
I only added the check because I thought it was be confusing for users to see they can connect a random unnamed model, one which they may not have even directly changed. For example, when a clip is added, two new models appear: "Mute", and "". It may appear odd to users adding a new clip and rightclicking on it to see that they can "Connect last changed model ()"

There are many other unnamed automatable models. It may be outside of the scope of this pr to give them all names, but for now I can remove the check for empty name just so that they function the same way. I suppose if they are visible, it will be easier for other people to track them down and name them eventually.

sakertooth commented on include/AutomatableModel.h:
```diff
@@ -94,6 +94,8 @@ class LMMS_EXPORT AutomatableModel : public Model, public JournallingObject
 	virtual void accept(ModelVisitor& v) = 0;
 	virtual void accept(ConstModelVisitor& v) const = 0;
 
+	static AutomatableModel* s_lastChangedModel;
```
I like to use `inline static` when I can to avoid the definition in the .cpp file. It's personal preference mostly.

```suggestion
inline static AutomatableModel* s_lastChangedModel = nullptr;
```

sakertooth commented on src/core/AutomatableModel.cpp:
```diff
@@ -37,6 +37,7 @@ namespace lmms
 {
 
 long AutomatableModel::s_periodCounter = 0;
+AutomatableModel* AutomatableModel::s_lastChangedModel = nullptr;
```
This could be removed if you use `inline static`. Up to you.

```suggestion
```

sakertooth reviewed (approved):
I am okay with this.

sakertooth commented:
Oops, I messed up the indent. I also think we can move `s_lastChangedModel` into the `private` section of the class? Sorry for the back and fourth.

regulus79 commented:
`AutomatableModel::s_lastChangedModel` is accessed by `AutomationClipView`, so if it is moved to private I believe we would need to add a getter method, but sure.

sakertooth commented:
> AutomatableModel::s_lastChangedModel is accessed by AutomationClipView, so if it is moved to private I believe we would need to add a getter method, but sure.

That's fine. Yeah, The getter method is nice because it only allows read access to the variable.

sakertooth commented:
Maybe I don't want to merge just yet: There's something I think I can do to remedy the unknown model name problem, at least when making clips.

regulus79 commented:
Also, has anyone tested this with a VST?

sakertooth commented:
> Also, has anyone tested this with a VST?

Automation with VST is unusable because of the spike in CPU, so there may not be a big use case for that at the moment.

> Maybe I don't want to merge just yet: There's something I think I can do to remedy the unknown model name problem, at least when making clips.

The problem is basically that when constructing automatable models, it calls `setValue` in the constructor. I was thinking of maybe moving `s_lastChangedModel` into `AutomatableModelView` somehow and overriding `mousePressEvent`, but come to realize that it doesn't even inherit `QWidget`. This idea might take some refactoring, which we probably don't want to do in this PR, so I'm fine with just saying "(Unknown model name)" instead of just "()"

BluesM18A1 commented:
> Also, has anyone tested this with a VST?

It seems like a few of my VSTs don't work here. Still investigating, will give details later.

sakertooth commented:
I'm having issues with this as `setValue` is called in the constructor, causing it to be set too soon. I also don't know how good of an idea is storing something like `s_lastChangedModel` in the Core when it only concerns the GUI layer.

We should stop storing `s_lastChangedModel` in `AutomatableModel`. It can probably go into `AutomatableModelView`. Just connect the `Model::dataChanged` signal to a lambda which sets `s_lastChangedModel`. That would be better than what is currently done now, IMO, but there's another problem.

Another problem I am seeing is if the project already has automation or LFO controllers attached to an `AutomatableModel`. It would just keep changing `s_lastChangedModel` when the user isn't changing it at all. We need some way to distinguish between `AutomatableModel` changes from the user and some other actor. A better idea maybe is to still store `s_lastChangedModel` in `AutomatableModelView`, but instead of taking the widget representing itself, it instead inherits `QWidget`. Then you can override `mousePressEvent` and set it there.

sakertooth commented:
Nevermind, it seems like the "consistently changing the model value" issue isn't there. I thought it was. Still don't like storing `s_lastChangedModel` in the Core though, but I guess this works for now..

regulus79 commented:
> We should stop storing s_lastChangedModel in AutomatableModel. It can probably go into AutomatableModelView. Just connect the Model::dataChanged signal to a lambda which sets s_lastChangedModel. That would be better than what is currently done now, IMO, but there's another problem.

Some ModelViews change their models after they have been initialized (for whatever reason), which makes this idea more complex. I can connect the model's `dataChanged` signal to a lambda to update `s_lastChangedModel`, but if the model changed via `AutomatableModelView::setModel`, then I will have to also connect that one to a lambda too. 

This is not necessarily an issue, but it makes the code a little less pretty since more checks have to be done and the connect statement appears in two places. It works, but it doesn't feel optimal, and it's easy to cause segfaults if I miss something.

sakertooth commented:
Could you give an example?

regulus79 commented:
When you add a new automation track (for example), it first creates the Mute and Solo models, but they appear to be temporary as they are soon replaced with new Mute and Solo models. (I suspect this may be due to a call to `Knob::setModel()` after the knob has been constructed.)

This is a small excerpt of me printing
- `qDebug() << this << "Connecting" << modelUntyped() << modelUntyped()->fullDisplayName() << "in AutomatableModelView constructor";` in the constructor and 
- `qDebug() << this << "AutomatableModelView::setModel was called with" << model << model->fullDisplayName() << "Old model:" << modelUntyped() << modelUntyped()->fullDisplayName();` in setModel()
```
0x5b122333c1a0 Connecting lmms::BoolModel(0x5b1223218cc0) "Mute" in AutomatableModelView constructor
0x5b1223210fb0 Connecting lmms::BoolModel(0x5b12232f60e0) "Solo" in AutomatableModelView constructor
0x5b122333c1a0 AutomatableModelView::setModel was called with lmms::BoolModel(0x77df3c005380) "Automation track>Mute" Old model: lmms::BoolModel(0x5b1223218cc0) "Mute"
0x5b1223210fb0 AutomatableModelView::setModel was called with lmms::BoolModel(0x77df3c005448) "Automation track>Solo" Old model: lmms::BoolModel(0x5b12232f60e0) "Solo"
```
Almost all models do this, so lmms is spamming debug info like this when it starts up.

sakertooth commented:
Got the program to crash by adding a mixer channel and then right-clicking on an automation clip. We might need to rethink the solution a bit.

regulus79 commented:
I cannot reproduce the crash. Did you add the mixer channel after adding the clip? Also did you change any knobs before doing it, or was it a fresh project?

sakertooth commented:
No, I just added a mixer channel with the "+" button, and then right clicked on an automation clip. I honestly don't know why you can't reproduce it. I'll try again soon.

regulus79 commented:
I can now reproduce the crash. It appears it only occurs when starting from an empty project.

firewall1110 commented:
Is `s_lastChangedModel`  always initialized properly?
May be some more` nullptr` checks needed ... 

P.S.
It is only quick impression - may be wrong ...

regulus79 commented:
I have moved the implementation back into the core, as having in the gui was causing some difficult to debug crashes.

I have also added initial support for VST knobs. Basically, after you move a knob in the VST and then press the VST sync button, the knobs which have changed from their previous values are updated using `AutomatableModel::setValue` which updates the last changed model. If the knobs have not been changed, they are updated with `AutomatableModel::setAutomatedValue` which does not update the last changed model. This is very sketchy, but maybe it will suffice until the VST messaging system is reworked?
